src/rateLimiters/slidingWindowLog.ts

```diff
@@ -50,9 +83,62 @@ class SlidingWindowLog implements RateLimiter {
     ): Promise<RateLimiterResponse> {
         // set the expiry of key-value pairs in the cache to 24 hours
         const keyExpiry = 86400000; // TODO: Make this a global for consistency across each algo.
-        if (tokens > this.capacity) return { success: false, tokens: this.capacity };
 
-        throw new Error('SlidingWindowLog.processRequest not implemented');
+        // Each user's log is represented by a redis list with a score = request timestamp
+        // and a value equal to the complexity
+        // Drop expired requests from the log. represented by a sorted set in redis
+
+        // Get the log from redis
+        let requestLog: RedisLog = JSON.parse((await this.client.get(uuid)) || '[]');
+
+        // Iterate through the list in reverse and count active tokens
+        // This allows us to track the threshold for when this request would be allowed if it is blocked
+        // Stop at the first timestamp that's expired and cut the rest.
+
+        const cutoff = timestamp - this.windowSize;
+        let tokensInLog = 0;
+        let cutoffIndex = 0; // index of oldest active request
+        // TODO: Provide a timestamp for when the request will succeeed.
+        // Compute time between response and this timestamp later on
+        // FIXME: What should this be if the complexity is too big?
+        let retryIndex = requestLog.length; // time the user must wait before a request can be allowed.
+
+        for (let index = requestLog.length - 1; index >= 0; index--) {
+            if (cutoff >= requestLog[index].timestamp) {
+                cutoffIndex = index + 1;
+                break;
+            } else {
+                // the request is active
+                tokensInLog += requestLog[index].tokens;
+                if (this.capacity - tokensInLog >= tokens) {
+                    // the log is able to accept this request
+                    retryIndex = index;
+                }
+            }
+        }
+
+        let retryAfter: number;
+        if (tokens > this.capacity) retryAfter = Infinity;
+        // need the request before retryIndex
+        else if (retryIndex > 0)
+            retryAfter = this.windowSize + requestLog[retryIndex - 1].timestamp;
```

Comment:
If `retryIndex` is the index in the log of the "blocking request" would you not want the `retryAfter` to be the timestamp of that request (+ 1ms) not the timestamp of the request before it?

Reply:
`retryIndex` was the index of the oldest request in the log when the current request would be accepted. So we would need to retryAfter the request at `retryIndex - 1` is removed. 

I'll rename the variables and add comments to clarify.

Reply:
The tests seem right so if it passes that's good.